### PR TITLE
Bug #84415 slave don't report Seconds_Behind_Master when running s lave_parallel_workers > 0

### PR DESCRIPTION
**Pull request by robgolebiowski**
The fix: When a task from GAQ finishes executing, last_master_timestamp gets assigned timestamp from next task in GAQ. In case GAQ is empty last_master_timestamp gets reset to 0. Because last_master_timestamp is only re-assigned when task leaves GAQ it must be also assigned when first event (from parallel execution) arrives.

jenkins run => http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/684/ - failing tests checked. Not failing locally or not failing due to the bug fix

### CODE REVIEW

File: sql/rpl_rli_pdb.h [hunk 1]
```diff
@@ -229,7 +229,12 @@ class circular_buffer_queue
   /**
      return the value of @c data member of the head of the queue.
   */
-  Element_type* head_queue();
+  Element_type* head_queue()
```
Comment (percona-ysorokin) [inline sql/rpl_rli_pdb.h:232]: Was there an old implementation in the `.cc` file? If so, it must be removed.

**Review by percona-ysorokin — COMMENTED** [review]
The same comments as in 5.6

File: mysql-test/suite/rpl/t/bug84415_logical_clock.test [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+SET @save_slave_parallel_workers = @@GLOBAL.slave_parallel_workers;
+SET @save_slave_transaction_retries = @@GLOBAL.slave_transaction_retries;
+
+SET GLOBAL slave_parallel_type = LOGICAL_CLOCK;
+SET GLOBAL slave_parallel_workers = 3;
+SET GLOBAL slave_transaction_retries = 0;
+
+--source include/start_slave.inc
+
+# 1st scenario. On slave - two statements are applied in parralel on one database.
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415_logical_clock.test]: "parallel" everywhere

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/suite/rpl/t/bug84415_logical_clock.test (hunk 2 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415_logical_clock.test]: Done.

File: sql/rpl_rli_pdb.h (hunk 1 above)
Comment (robgolebiowski) [reply sql/rpl_rli_pdb.h:232]: No, there was only declaration and no implementation.

**Comment by robgolebiowski** [comment]
Jenkins run => http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/717/

File: mysql-test/suite/rpl/t/bug84415_logical_clock.test [hunk 3]
```diff
@@ -0,0 +1,357 @@
+--let $rpl_skip_start_slave= 1
+--source include/master-slave.inc
+# Must be statement as we relay on PROCESSLIST's Info column
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/bug84415_logical_clock.test]: s/relay/rely

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Same minor comments as for 5.6 + one typo

File: mysql-test/suite/rpl/t/bug84415_logical_clock.test [hunk 4]
```diff
@@ -0,0 +1,357 @@
+--let $rpl_skip_start_slave= 1
+--source include/master-slave.inc
+# Must be statement as we rely on PROCESSLIST's Info column
+--source include/have_binlog_format_statement.inc
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/bug84415_logical_clock.test]: `have_*` includes before `master-slave`

File: mysql-test/suite/rpl/t/bug84415_logical_clock.test [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+##
+# cleanup
+##
+connection master;
+DROP TABLE test.t1;
+DROP TABLE test.t2;
+DROP DATABASE second_test;
+
+connection slave;
+#DROP USER test;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/bug84415_logical_clock.test:348]: please remove

File: mysql-test/suite/rpl/t/bug84415-slave.opt [hunk 6]
```diff
@@ -0,0 +1 @@
+--slave-transaction-retries=0
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/bug84415-slave.opt]: is there a way to set `slave-transaction-retries` dynamically to get rid of `.opt` file?

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]
Just minor comments

File: mysql-test/suite/rpl/t/bug84415_logical_clock.test (hunk 4 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415_logical_clock.test]: Done.

File: mysql-test/suite/rpl/t/bug84415_logical_clock.test (hunk 5 above)
Comment (robgolebiowski) [reply mysql-test/suite/rpl/t/bug84415_logical_clock.test:348]: Done

**Comment by robgolebiowski** [comment]
Jenkins run => http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/725/

**Review by laurynas-biveinis — APPROVED** [review]

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request merged**